spec/lib/simple_token_authentication/token_authentication_handler_spec.rb

```diff
@@ -693,29 +693,47 @@
   end
 
   describe '#authenticate_entity_from_token!' do
-
     let(:token_authentication_handler) { described_class.new }
 
     before(:each) do
-        allow(token_authentication_handler).to receive(:find_record_from_identifier)
-        allow(token_authentication_handler).to receive(:perform_sign_in!)
-        allow(token_authentication_handler).to receive(:token_correct?).and_return(false)
+      allow(token_authentication_handler).to receive(:find_record_from_identifier)
+      allow(token_authentication_handler).to receive(:perform_sign_in!)
+      allow(token_authentication_handler).to receive(:token_correct?).and_return(false)
     end
 
-    it 'does not trigger the :after_successful_token_authentication hook', hooks: true, private: true do
-      expect(token_authentication_handler).not_to receive(:after_successful_token_authentication)
-      token_authentication_handler.send(:authenticate_entity_from_token!, double)
-    end
+    context 'when authentication is not succesful and the handler implements the :after_successful_token_authentication hook', private: true do
+      before(:each) do
+        allow(token_authentication_handler).to receive(:token_correct?).and_return(false)
+        allow(token_authentication_handler).to receive(:after_successful_token_authentication)
```

Comment:
This indeed looks very good to me.

These tests are very mock-heavy, which is probably a smell in itself, but in that context, I think you're making good decisions in the way you re-wrote the specs. :100: